RealmSwift/Tests/RealmTests.swift

```diff
@@ -497,92 +497,91 @@ class RealmTests: TestCase {
+        testIntPrimaryKey(for: SwiftPrimaryInt64Object.self)
+    }
 
-            // When this is directly inside the XCTAssertNil, it fails for some reason
-            let missingObject = realm.object(ofType: SwiftPrimaryStringObject.self, forPrimaryKey: "z")
-            XCTAssertNil(missingObject)
-        }
+    func testOptionalIntPrimaryKey() {
+        func testOptionalIntPrimaryKey<O: Object, Wrapped: RealmOptionalType>(for type: O.Type)
+            where O: SwiftPrimaryKeyObjectType, O.PrimaryKey == RealmOptional<Wrapped>,
+                  Wrapped: ExpressibleByIntegerLiteral {
+                let realm = try! Realm()
+                try! realm.write {
+                    realm.createObject(ofType: type, populatedWith: ["a", NSNull()])
+                    realm.createObject(ofType: type, populatedWith: ["b", 2])
+                }
 
-        do {
-            let object1 = realm.object(ofType: SwiftPrimaryOptionalStringObject.self, forPrimaryKey: NSNull())
-            XCTAssertNotNil(object1)
+                let object1 = realm.object(ofType: type, forPrimaryKey: NSNull())
+                XCTAssertNotNil(object1)
 
-            let object2 = realm.object(ofType: SwiftPrimaryOptionalStringObject.self, forPrimaryKey: "b")
-            XCTAssertNotNil(object2)
+                let object2 = realm.object(ofType: type, forPrimaryKey: 2 as Wrapped)
+                XCTAssertNotNil(object2)
 
-            let missingObject = realm.object(ofType: SwiftPrimaryOptionalStringObject.self, forPrimaryKey: "z")
-            XCTAssertNil(missingObject)
+                let missingObject = realm.object(ofType: type, forPrimaryKey: 0 as Wrapped)
+                XCTAssertNil(missingObject)
         }
 
-        func assertIntObject(_ objectType: Object.Type) {
-            let object = realm.object(ofType: objectType, forPrimaryKey: 1)
-            XCTAssertNotNil(object)
+        testOptionalIntPrimaryKey(for: SwiftPrimaryOptionalIntObject.self)
+        testOptionalIntPrimaryKey(for: SwiftPrimaryOptionalInt8Object.self)
+        testOptionalIntPrimaryKey(for: SwiftPrimaryOptionalInt16Object.self)
+        testOptionalIntPrimaryKey(for: SwiftPrimaryOptionalInt32Object.self)
+        testOptionalIntPrimaryKey(for: SwiftPrimaryOptionalInt64Object.self)
+    }
 
-            let missingObject = realm.object(ofType: objectType, forPrimaryKey: 0)
-            XCTAssertNil(missingObject)
+    func testStringPrimaryKey() {
+        let realm = try! Realm()
+        try! realm.write {
+            realm.createObject(ofType: SwiftPrimaryStringObject.self, populatedWith: ["a", 1])
+            realm.createObject(ofType: SwiftPrimaryStringObject.self, populatedWith: ["b", 2])
         }
 
-        func assertOptionalIntObject(_ objectType: Object.Type) {
-            let object1 = realm.object(ofType: objectType, forPrimaryKey: NSNull())
-            XCTAssertNotNil(object1)
+        // When this is directly inside the XCTAssertNotNil, it doesn't work
```

Comment:
I didn't introduce this comment, so no clue. I could test and see if this is still and issue. Maybe we can get rid of it.